### PR TITLE
Add mechanism to calculate average cost

### PR DESCRIPTION
**Pull request by vessla**
Kilka zaznaczonych zmian dodało niestety automatyczne formatowanie :(

### CODE REVIEW

File: src/main/java/org/opentripplanner/hasura_client/VehiclePositionsGetter.java [hunk 1]
```diff
@@ -9,11 +9,15 @@
...
 import org.slf4j.LoggerFactory;
 
+import java.util.Collections;
+import java.util.Objects;
 import java.util.Set;
 
 public class VehiclePositionsGetter extends HasuraGetter<VehicleDescription, Vehicle> {
     private static final Logger LOG = LoggerFactory.getLogger(VehiclePositionsGetter.class);
 
+    private VehiclePositionsMapper mapper;
```
Comment (MironSz) [inline src/main/java/org/opentripplanner/hasura_client/VehiclePositionsGetter.java]: Those are only style changes, right?

File: src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedVehiclesUpdater.java [hunk 2]
```diff
@@ -31,6 +31,14 @@ protected void runPolling() {
                 vehiclePositionsGetter.getResponsiveProviders()));
     }
 
+    public void readFromSnapshot() {
```
Comment (MironSz) [inline src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicles_positions/SharedVehiclesUpdater.java:34]: Shouldn't those changes be in separate PR?

File: src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/hasura_client/mappers/VehicleStateSnapshotMapper.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    protected VehicleDescription mapSingleHasuraObject(VehicleStateSnapshotHasuraObject vehicleStateSnapshotHasuraObject) {
+        VehiclePositionsMapper originalVehicleMapper = new VehiclePositionsMapper();
+        Provider vehicleProvider = vehicleProviders.get(vehicleStateSnapshotHasuraObject.getProviderId());
+        if (Objects.nonNull(vehicleProvider)) {
+            /*This is a little bit ugly workaround to automatically incorporate all future modifications of the original
+            vehicle mapping method code here, without changing the existing vehicle mapper interface*/
+            Vehicle stateSnapshotAsVehicle = vehicleStateSnapshotHasuraObject.toVehicle(vehicleProvider);
+            List<VehicleDescription> mappedVehicles = originalVehicleMapper.map(Arrays.asList(stateSnapshotAsVehicle));
```
Comment (AdamWiktor) [inline src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/hasura_client/mappers/VehicleStateSnapshotMapper.java]: `singletonList(stateSnapshotAsVehicle)`

File: src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/DatabaseSnapshotDownloader.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class DatabaseSnapshotDownloader {
+
+    private static final String DEFAULT_SNAPSHOT_FILE_NAME = "current_snapshot.json";
+
+    private Graph graph;
```
Comment (AdamWiktor) [inline src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/DatabaseSnapshotDownloader.java]: `final` on all fields

File: src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/RandomLocationGenerator.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+public class RandomLocationGenerator {
+
+    private Random random;
+
+    public RandomLocationGenerator(Random random) {
+        this.random = random;
+    }
+
+    public GenericLocation generateRandomLocation(GenericLocation center, double radius) {
```
Comment (AdamWiktor) [inline src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/RandomLocationGenerator.java:17]: I am not 100% sure, but I think this code doesn't generate random location with even distribution

Comment (AdamWiktor) [reply src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/RandomLocationGenerator.java:17]: I looked it up and it is actually a proper algorithm - nice!

File: src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/EstimatorCommandLineParameters.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    @Parameter(names = {"--snapshotInterval"}, required = true, validateWith = ValidPositiveInteger.class, description = "Interval for snapshots (in minutes)")
+    private int snapshotIntervalInMinutes;
+
+    @Parameter(names = {"--snapshotDatabaseURL"}, required = true, description = "URL from which we want to download snapshots")
+    private String databaseURL;
+
+    @Parameter(names = {"--snapshotDatabasePass"}, required = true, description = "URL from which we want to download snapshots")
+    private String databasePassword;
+
+    public String getRouterName() {
```
Comment (AdamWiktor) [inline src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/EstimatorCommandLineParameters.java]: `@Getter` on a class

File: src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/MobilityPackagePriceEstimator.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.List;
+import java.util.Random;
+
+import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.options;
+
+public class MobilityPackagePriceEstimator {
```
Comment (AdamWiktor) [inline src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/MobilityPackagePriceEstimator.java:33]: Wow, so thats how python scripts would look if they were written in java xD

File: src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/DatabaseSnapshotDownloader.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        List<VehicleDescription> vehicles = vehicleSnapshotGetter.postFromHasuraWithPassword(this.graph, this.databaseURL, this.databasePassword);
+        if (Objects.nonNull(vehicles)) {
+            saveSnapshotData(vehicles);
+            return vehicles.size();
+        }
+        return 0;
+    }
+
+    private void saveSnapshotData(List<VehicleDescription> vehicles) {
+        try (BufferedWriter fileWriter = new BufferedWriter(new FileWriter(snapshotDirectory + DEFAULT_SNAPSHOT_FILE_NAME))) {
```
Comment (AdamWiktor) [inline src/mobilityPackagePriceEstimation/java/org/opentripplanner/estimator/utils/DatabaseSnapshotDownloader.java]: Ths code begs to be replaced by `@Value` classes and some json serialization instead of doing it by hand

**Review by AdamWiktor — APPROVED** [review]

File: src/main/java/org/opentripplanner/hasura_client/VehiclePositionsGetter.java (hunk 1 above)
Comment (vessla) [reply src/main/java/org/opentripplanner/hasura_client/VehiclePositionsGetter.java]: I have compared it to the dev-1.x branch locally and the only difference was the removal of an empty line above the "private static final Logger LOG ..." initialization
